assets/js/ui/file_download.ts

```diff
@@ -0,0 +1,71 @@
+import showFlashMessage from "./flash_message";
+
+export default function fileDownload(el) {
+  el.querySelectorAll("a[download]").forEach((anchor) => {
+    anchor.addEventListener("click", async (e) => {
+      e.preventDefault();
+
+      const flashMessage = showFlashMessage({
+        isLoading: true,
+        text: "Preparing download...",
+        isDismissible: false,
+        toastOptions: {
+          autohide: false,
+        },
+      });
+
+      try {
+        const response = await fetch(anchor.href);
+        if (!response.ok) {
+          throw new Error(
+            "An unexpected error occurred while downloading the file. Please try again.",
+          );
+        }
+
+        const filename = extractFileName(response);
+        const blob = await response.blob();
+
+        const dummyAnchor = document.createElement("a");
+        dummyAnchor.href = URL.createObjectURL(blob);
+        dummyAnchor.setAttribute("class", "link-primary text-nowrap");
+        dummyAnchor.setAttribute("download", filename);
+        dummyAnchor.textContent = filename;
+
+        flashMessage?.setLevel("success");
+        flashMessage?.setIsLoading(false);
+        flashMessage?.setText("Download ready: " + dummyAnchor.outerHTML);
+        flashMessage?.setIsDismissible(true);
+
+        // Trigger download (save-as window or auto-download, depending on browser settings)
+        dummyAnchor.click();
```

Comment:
This is actually a hack that's used widespread. The click on the `<a>` tag is prevented at the top of this function so we can intervene and display the toast first. Then the file is downloaded via the `fetch()` API. When ready, we create a dummy anchor tag using a data URL with the downloaded file. At the end, the `click()` method is triggered on that which will actually present the download to the browser. The dummy anchor is also included in the updated toast, but that is not necessary. Just an alternate way to get the file, in case you canceled the save-as-dialog.

Tested in Chrome, Firefox, Safari and Edge.